File: apps/backend/app/infrastructure/knowledge/mappers/document_mapper.py
from app.domain.knowledge.entities.document import Document
from app.domain.knowledge.enums.document_status import DocumentStatus
from app.infrastructure.knowledge.models.document_model import (
    DocumentModel,
)


def to_domain(
    model: DocumentModel,
) -> Document:
    """
    Maps a SQLAlchemy DocumentModel to a domain Document.
    """

    return Document(
        id=model.id,
        owner_id=model.owner_id,
        original_filename=model.original_filename,
        content_type=model.content_type,
        size_bytes=model.size_bytes,
        storage_key=model.storage_key,
        status=DocumentStatus(model.status),
        created_at=model.created_at,
        updated_at=model.updated_at,
    )


def to_model(
    entity: Document,
) -> DocumentModel:
    """
    Maps a domain Document to a SQLAlchemy DocumentModel.
    """

    return DocumentModel(
        id=entity.id,
        owner_id=entity.owner_id,
        original_filename=entity.original_filename,
        content_type=entity.content_type,
        size_bytes=entity.size_bytes,
        storage_key=entity.storage_key,
        status=entity.status.value,
        created_at=entity.created_at,
        updated_at=entity.updated_at,
    )

File: apps/frontend/src/app/data/mappers/document.mapper.ts
import { DocumentDto, IndexDocumentResponseDto } from '@data/models/document.dto';
import { DocumentStatus, KnowledgeDocument } from '@features/documents/models/knowledge-document.model';
import { IndexResult } from '@features/documents/models/index-result.model';

export function mapDocumentDtoToModel(dto: DocumentDto): KnowledgeDocument {
  return {
    id: dto.id,
    filename: dto.filename,
    contentType: dto.content_type,
    sizeBytes: dto.size_bytes,
    status: normalizeStatus(dto.status),
    createdAt: new Date(dto.created_at),
    updatedAt: dto.updated_at ? new Date(dto.updated_at) : new Date(dto.created_at)
  };
}

export function mapIndexResponseDtoToModel(dto: IndexDocumentResponseDto): IndexResult {
  return {
    documentId: dto.document_id,
    status: normalizeStatus(dto.status),
    chunkCount: dto.chunk_count
  };
}

/**
 * Normalizes case defensively (see the DTO's status field comment)
 * and fails safe: an unrecognized status maps to 'failed' rather than
 * 'available'/'indexed', so the UI never implies a document is in a
 * better state than we can actually confirm from what the backend sent.
 */
function normalizeStatus(raw: string): DocumentStatus {
  const normalized = raw.toLowerCase();
  if (normalized === 'available' || normalized === 'indexing' || normalized === 'indexed' || normalized === 'failed') {
    return normalized;
  }
  return 'failed';
}
